Faulty implementation of Server#getRegistry(Class)

### Expected behavior

Per [Bukkit Javadocs](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/Server.html#getRegistry(java.lang.Class)), the Server#getRegistry(Class) method returns null when a registry is not available for the given Keyed class.

### Observed/Actual behavior

Paper asserts that a registry _must_ exists and throws a NullPointerException:

```
[10:40:39 ERROR]: Exception when Jannyboy11 attempted to tab complete Jannyboy11
org.bukkit.command.CommandException: Unhandled exception during tab completion for command '/igive Jannyboy11 ' in plugin InvSeePlusPlus_Give v0.29.6-SNAPSHOT
        at org.bukkit.command.PluginCommand.tabComplete(PluginCommand.java:150) ~[paper-mojangapi-1.21.1-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.command.Command.tabComplete(Command.java:101) ~[paper-mojangapi-1.21.1-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.command.brigadier.bukkit.BukkitCommandNode$BukkitBrigSuggestionProvider.getSuggestions(BukkitCommandNode.java:118) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at com.mojang.brigadier.tree.ArgumentCommandNode.listSuggestions(ArgumentCommandNode.java:71) ~[brigadier-1.3.10.jar:1.21.1-DEV-925c3b9]
        at com.mojang.brigadier.CommandDispatcher.getCompletionSuggestions(CommandDispatcher.java:551) ~[paper-1.21.1.jar:?]
        at com.mojang.brigadier.CommandDispatcher.getCompletionSuggestions(CommandDispatcher.java:531) ~[paper-1.21.1.jar:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.sendServerSuggestions(ServerGamePacketListenerImpl.java:870) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.lambda$handleCustomCommandSuggestions0$2(ServerGamePacketListenerImpl.java:831) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.TickTask.run(TickTask.java:18) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:151) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1537) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:201) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:125) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1514) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1507) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:135) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.MinecraftServer.managedBlock(MinecraftServer.java:1466) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:1473) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1318) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:330) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.NullPointerException: class org.bukkit.Material is not a valid registry type
        at java.base/java.util.Objects.requireNonNull(Objects.java:360) ~[?:?]
        at io.papermc.paper.registry.PaperRegistryAccess.getRegistry(PaperRegistryAccess.java:48) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at org.bukkit.craftbukkit.CraftServer.getRegistry(CraftServer.java:2946) ~[paper-1.21.1.jar:1.21.1-DEV-925c3b9]
        at org.bukkit.Bukkit.getRegistry(Bukkit.java:2588) ~[paper-mojangapi-1.21.1-R0.1-SNAPSHOT.jar:?]
        at InvSee++.jar/com.janboerman.invsee.spigot.impl_1_21_1_R1.InvseeImpl.materials(InvseeImpl.java:344) ~[InvSee++.jar:?]
        at InvSee++.jar/com.janboerman.invsee.spigot.InvseePlusPlus.materials(InvseePlusPlus.java:221) ~[InvSee++.jar:?]
        at InvSee++_Give.jar/com.janboerman.invsee.spigot.addon.give.GiveTabCompleter.onTabComplete(GiveTabCompleter.java:47) ~[InvSee++_Give.jar:?]
        at org.bukkit.command.PluginCommand.tabComplete(PluginCommand.java:138) ~[paper-mojangapi-1.21.1-R0.1-SNAPSHOT.jar:?]
        ... 21 more
```

### Steps/models to reproduce

1. Install my plugins InvSee++ and InvSee++_Give: https://hangar.papermc.io/Jannyboy11/InvSee-plus-plus
2. Use the command `/invgive Notch ` and then press tab to tabcomplete materials.

### Plugin and Datapack List

Plugins: InvSeePlusPlus, InvSeePlusPlus_Give

No datapacks were used.

### Paper version

 This server is running Paper version 1.21.1-DEV-master@925c3b9 (2024-09-08T17:37:37Z) (Implementing API version 1.21.1-R0.1-SNAPSHOT)

I know this is running 16 versions behind, but the bug seems to be present still, per this patch: https://github.com/PaperMC/Paper/blob/master/patches/server/0471-Add-RegistryAccess-for-managing-Registries.patch#L878

### Other

Per [this message](https://discord.com/channels/289587909051416579/1077385604012179486/1263418959554805843) on the PaperMC discord, I implemented this new method to obtain the set of available materials on the server, as can be seen in the source code of my plugin here (using the 'upstream compatible' api): https://github.com/Jannyboy11/InvSee-plus-plus/blob/5c09d698fa74d302ae88d50dcdc02a79036b0735/InvSee%2B%2B_Platforms/Impl_1_21_1_R1/src/main/java/com/janboerman/invsee/spigot/impl_1_21_1_R1/InvseeImpl.java#L344

If this is the incorrect way to do it, then I'd like to know.

--- a/patches/server/0471-Add-RegistryAccess-for-managing-Registries.patch
+++ b/patches/server/0471-Add-RegistryAccess-for-managing-Registries.patch
@@ -164,10 +164,10 @@ index 0000000000000000000000000000000000000000..70e2c3b5cac9a0dfb043de218df20dc1
 +}
 diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistryAccess.java b/src/main/java/io/papermc/paper/registry/PaperRegistryAccess.java
 new file mode 100644
-index 0000000000000000000000000000000000000000..d591e3a2e19d5358a0d25a5a681368943622d231
+index 0000000000000000000000000000000000000000..35b6a7c5bac9640332a833bd3627f2bcb1bbf2f3
 --- /dev/null
 +++ b/src/main/java/io/papermc/paper/registry/PaperRegistryAccess.java
-@@ -0,0 +1,124 @@
+@@ -0,0 +1,127 @@
 +package io.papermc.paper.registry;
 +
 +import io.papermc.paper.registry.entry.ApiRegistryEntry;
@@ -213,10 +213,13 @@ index 0000000000000000000000000000000000000000..d591e3a2e19d5358a0d25a5a68136894
 +    @Deprecated(forRemoval = true)
 +    @Override
 +    public <T extends Keyed> @Nullable Registry<T> getRegistry(final Class<T> type) {
-+        final RegistryKey<T> registryKey;
-+        final @Nullable RegistryEntry<?, T> entry;
-+        registryKey = requireNonNull(byType(type), () -> type + " is not a valid registry type");
-+        entry = PaperRegistries.getEntry(registryKey);
++        final @Nullable RegistryKey<T> registryKey = byType(type);
++        // If our mapping from Class -> RegistryKey did not contain the passed type it was either a completely invalid type or a registry
++        // that merely exists as a SimpleRegistry in the org.bukkit.Registry type. We cannot return a registry for these, return null
++        // as per method contract in Bukkit#getRegistry.
++        if (registryKey == null) return null;
++
++        final @Nullable RegistryEntry<?, T> entry = PaperRegistries.getEntry(registryKey);
 +        final @Nullable RegistryHolder<T> registry = (RegistryHolder<T>) this.registries.get(registryKey);
 +        if (registry != null) {
 +            // if the registry exists, return right away. Since this is the "legacy" method, we return DelayedRegistry

--- a/patches/server/1015-Registry-Modification-API.patch
+++ b/patches/server/1015-Registry-Modification-API.patch
@@ -45,10 +45,10 @@ index 70e2c3b5cac9a0dfb043de218df20dc1ab2cc070..38222dca2497cec5b104b21429a9ec3a
      }
  }
 diff --git a/src/main/java/io/papermc/paper/registry/PaperRegistryAccess.java b/src/main/java/io/papermc/paper/registry/PaperRegistryAccess.java
-index d591e3a2e19d5358a0d25a5a681368943622d231..f05ebf453406a924da3de6fb250f4793a1b3c612 100644
+index 35b6a7c5bac9640332a833bd3627f2bcb1bbf2f3..1026b9c04f94ed73049b980822a2eafdbacea7fd 100644
 --- a/src/main/java/io/papermc/paper/registry/PaperRegistryAccess.java
 +++ b/src/main/java/io/papermc/paper/registry/PaperRegistryAccess.java
-@@ -80,6 +80,14 @@ public class PaperRegistryAccess implements RegistryAccess {
+@@ -83,6 +83,14 @@ public class PaperRegistryAccess implements RegistryAccess {
          return possiblyUnwrap(registryHolder.get());
      }
  

